timelock-api/src/main/java/com/palantir/atlasdb/timelock/watch/ExplicitLockPredicate.java

```diff
@@ -0,0 +1,39 @@
+/*
+ * (c) Copyright 2019 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.timelock.watch;
+
+import java.util.Set;
+
+import org.immutables.value.Value;
+
+import com.fasterxml.jackson.annotation.JsonTypeName;
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import com.fasterxml.jackson.databind.annotation.JsonSerialize;
+import com.palantir.lock.LockDescriptor;
+
+@JsonSerialize(as = ImmutableExplicitLockPredicate.class)
+@JsonDeserialize(as = ImmutableExplicitLockPredicate.class)
+@Value.Immutable
+@JsonTypeName("explicit")
+public interface ExplicitLockPredicate extends LockPredicate {
+    Set<LockDescriptor> descriptors();
```

Comment:
Will be interesting to think about how broadly we end up applying this, as it scales Timelock from only retaining the collection of locks for ongoing writes to retaining some extra amount of objects that's theoretically unrelated to the write load.

The user also presumably has to pick the right one of row/cell locking corresponding to the table in question.

Reply:
Yeah, that's why the impl has a limit on the number of watches you can register. I should probably limit the number of descriptors too.

Correct - users who want to cache things need to know the type of locking involved (but I would imagine they would as they have the schema).